### PR TITLE
fix: workers creation race

### PR DESCRIPTION
**Pull request by dgozman**
Consider the following scenario: we have two tests that should run in different worker.

Currently, the following race is possible:
1. We create a worker for the first test and run it.
2. We need a worker for the second test, and since first worker is busy, we spawn a new one.
3. While second worker is initializing, the first one finishes and is ready for reuse.
4. We grab the first worker for the second test. However, it does not match the hash, so we spawn the third worker.

Overall, we get three workers instead of two. The solution is to specifically wait for the newly created worker, and leave reusable workers for future tests.

This race was often seen on the bots when running "should work with two different fixture objects".

### CODE REVIEW

File: src/dispatcher.ts [hunk 1]
```diff
@@ -150,9 +150,11 @@ export class Dispatcher {
       const requiredHash = file.hash;
       let worker = await this._obtainWorker();
       while (!this._isStopped && worker.hash && worker.hash !== requiredHash) {
-        this._restartWorker(worker);
+        worker.stop();
```
Comment (pavelfeldman) [inline src/dispatcher.ts:153]: To make sure I understand, for 3 workers, 3 tests with hash "A" and 3 tests with hash "B", we create 3 workers for the first 3 tests. Then we wait for them to become available. We get three workers back and we schedule 'stop' for them. Then we start obtaining the new one. There are no free workers and `this._workers.size` is still 3, so we become a claimer. But who is going to create a new worker?

File: src/dispatcher.ts [hunk 2]
```diff
@@ -216,23 +219,30 @@ export class Dispatcher {
...
-    // If there is worker, use it.
-    if (this._freeWorkers.length)
-      return this._freeWorkers.pop();
-    // If we can create worker, create it.
-    if (this._workers.size < this._config.workers)
-      this._createWorker();
-    // Wait for the next available worker.
-    await new Promise(f => this._workerClaimers.push(f));
-    return this._freeWorkers.pop();
+    const claimWorker = async () => {
```
Comment (pavelfeldman) [inline src/dispatcher.ts:222]: drop 'async'

File: src/dispatcher.ts [hunk 3]
```diff
@@ -216,23 +219,30 @@ export class Dispatcher {
...
+      // Use available worker.
+      if (this._freeWorkers.length)
+        return this._freeWorkers.pop();
+      // Create a new worker.
+      if (this._workers.size < this._config.workers)
+        return this._createWorker();
+      return null;
+    };
+
+    let worker = await claimWorker();
```
Comment (pavelfeldman) [inline src/dispatcher.ts:232]: Drop await, otherwise you could get _notifyWorkerClaimer called while you are in a microtask.

**Review by pavelfeldman — APPROVED** [review]

File: src/dispatcher.ts (hunk 1 above)
Comment (dgozman) [reply src/dispatcher.ts:153]: Fixed!

**Review by dgozman — COMMENTED** [review]
All fixed.

**Pull request merged**